pkg/hostap/monitor.go

```diff
@@ -0,0 +1,239 @@
+	hostapif "github.com/openshift-kni/eapol-operator/pkg/netlink"
+)
+
+const (
+	staConnectedEvent    = "AP-STA-CONNECTED"
+	eapSuccessEvent      = "CTRL-EVENT-EAP-SUCCESS"
+	staDisconnectedEvent = "AP-STA-DISCONNECTED"
+	eapFailureEvent      = "CTRL-EVENT-EAP-FAILURE"
+	pingCommand          = "PING"
+	attachCommand        = "ATTACH"
+	unixDgramProtocol    = "unixgram"
+	hostapdSocketDir     = "/var/run/hostapd/"
+	sockReadBufSize      = 4096
+)
+
+type InterfaceMonitor struct {
+	Logger            log.Logger
+	IfName            string
+	IfEventHandler    hostapif.LinkEventHandler
+	ifEventCh         chan netlink.LinkUpdate
+	hostApdConn       net.Conn
+	authenticatedMacs map[string]interface{}
+	mutex             sync.Mutex
+	stopWg            sync.WaitGroup
+	stop              chan interface{}
+	operState         netlink.LinkOperState
+}
+
+func (m *InterfaceMonitor) StartMonitor() error {
+	local, err := ioutil.TempFile(hostapdSocketDir, "hostapd_monitor_client")
+	if err != nil {
+		return err
+	}
+	os.Remove(local.Name())
+	conn, err := net.DialUnix(unixDgramProtocol, &net.UnixAddr{Name: local.Name(), Net: unixDgramProtocol},
+		&net.UnixAddr{Name: filepath.Join(hostapdSocketDir, m.IfName), Net: unixDgramProtocol})
+	if err != nil {
+		return err
+	}
+	m.hostApdConn = conn
+	m.authenticatedMacs = make(map[string]interface{})
+	m.stop = make(chan interface{})
+	m.stopWg.Add(3)
+	m.ifEventCh = make(chan netlink.LinkUpdate)
+	m.IfEventHandler.Subscribe(m.IfName, m.ifEventCh)
+	go m.handleHostapdReply()
+	go m.handleIfEvents()
+	go m.sendKeepAlive()
+	err = m.attachHostapd()
+	if err != nil {
+		m.StopMonitor()
+		return err
+	}
+	level.Info(m.Logger).Log("op", "monitor", "interface monitor started", m.IfName)
+	return nil
+}
+
+func (m *InterfaceMonitor) StopMonitor() {
+	close(m.stop)
+	m.IfEventHandler.Unsubscribe(m.IfName)
+	err := m.hostApdConn.Close()
+	if err != nil {
+		level.Error(m.Logger).Log("sockread", "error closing connection", m.IfName, err)
+	}
+	m.stopWg.Wait()
+}
+
+func (m *InterfaceMonitor) handleHostapdReply() {
+	defer m.stopWg.Done()
+	receivedByteArr := make([]byte, sockReadBufSize)
+	for {
+		select {
+		case <-m.stop:
+			return
+		default:
+			m.hostApdConn.SetReadDeadline(time.Now().Add(1 * time.Second))
+			size, err := m.hostApdConn.Read(receivedByteArr)
+			if err != nil {
+				if opErr, ok := err.(*net.OpError); ok && opErr.Timeout() {
+					continue
+				}
+				level.Error(m.Logger).Log("sockread", "error reading from connection", m.IfName, err)
+				return
+			}
+			if size == 0 {
+				continue
+			}
+			eventStr := string(receivedByteArr[:size])
+			err = m.handleHostapdEvent(eventStr)
+			if err != nil {
+				level.Error(m.Logger).Log("hostapd-event", eventStr, "interface", m.IfName, "error", err)
+			}
+		}
+	}
+}
+
+func (m *InterfaceMonitor) handleHostapdEvent(eventStr string) error {
+	eventStrSlice := strings.Split(eventStr, " ")
+	if len(eventStrSlice) < 2 {
```

Comment:
Probably we should log unexpectedly short messages?  Or do we already log all messages?